### PR TITLE
Recommendation system Step 1 & 2

### PR DESCRIPTION
**Pull request by Iamogeee**
### Complete implementation of step 1 and 2 of the recommendation system. 
Data Preprocessing Enhancements:

- Implemented detailed preprocessing steps in dataPreprocessor.js to handle missing values, normalize text data, one-hot encode categorical data, and apply TF-IDF encoding to text data.
- Ensured consistency in feature extraction and addressed any shape mismatches in the data.

Model Training and Saving:

- Created and executed trainModel.js to train a machine learning model using TensorFlow.js.
- Added validation and debugging steps to ensure feature vectors and labels are consistent in shape and format.
- Successfully saved the trained model to the backend/model directory.

### Added the following components:
ReviewForm 
FavoriteBusinesses
Modal
SideBar

<div>
    <a href="https://www.loom.com/share/32a1373e7521484ab95ce09f9d176ab6">
      <p>Loom Message - 11 July 2024 - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/32a1373e7521484ab95ce09f9d176ab6">
      
    </a>
  </div>

### CODE REVIEW

File: frontend/src/Components/BusinessCard/BusinessCard.jsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import React from "react";
+import { useNavigate } from "react-router-dom";
+import "./BusinessCard.css";
+
+const BusinessCard = ({ business, apiKey, onLike, onSave, onClick }) => {
+  const navigate = useNavigate();
+
+  const handleDetailsClick = () => {
+    navigate(`/business/${business.id}`);
+    onClick(business.id); // Tracking clicks/views
```
Comment (sinap-fb) [inline frontend/src/Components/BusinessCard/BusinessCard.jsx:10]: nice comment 👍

File: frontend/src/Components/BusinessDetails/BusinessDetails.jsx [hunk 2]
```diff
@@ -5,21 +5,44 @@ import BusinessOverview from "../BusinessOverview/BusinessOverview";
...
 
   useEffect(() => {
     // Fetching the details of the selected business from my backend
-    fetch(`http://localhost:3000/api/businesses/${id}`)
-      .then((response) => response.json())
-      .then((data) => setBusiness(data));
+    const fetchBusinessData = async () => {
+      try {
+        const response = await fetch(
+          `http://localhost:3000/api/businesses/${id}`
```
Comment (sinap-fb) [inline frontend/src/Components/BusinessDetails/BusinessDetails.jsx:26]: ~~does the `http://localhost:3000` part need to be here? i don't think this will work once deployed~~
nevermind, just saw the comment in the previous PR

File: frontend/src/Components/BusinessOverview/BusinessOverview.jsx [hunk 3]
```diff
@@ -5,7 +5,7 @@ const BusinessOverview = ({ overview }) => {
   return (
     <div className="business-overview">
       <h2>Overview</h2>
-      <p>{overview}</p>
+      <p>{overview || "Overview not provided"}</p>
```
Comment (sinap-fb) [inline frontend/src/Components/BusinessOverview/BusinessOverview.jsx:8]: 👌

File: frontend/src/index.css [hunk 4]
```diff
@@ -12,7 +12,11 @@
   -webkit-font-smoothing: antialiased;
   -moz-osx-font-smoothing: grayscale;
 }
-
+#root {
+  width: 100%;
+  min-height: 100vh;
+  padding-bottom: 10vh;
```
Comment (sinap-fb) [inline frontend/src/index.css:18]: what's this padding for? might be worth a comment

File: frontend/src/Components/SideBar/SideBar.jsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  const [isOpen, setIsOpen] = useState(false);
+
+  const toggleSidebar = () => {
+    setIsOpen(!isOpen);
+  };
+
+  return (
+    <>
+      <button className="menu-btn" onClick={toggleSidebar}>
+        ☰
```
Comment (sinap-fb) [inline frontend/src/Components/SideBar/SideBar.jsx:14]: this  button (as well as the X button) should probably have `aria-label`s since screen readers would otherwise probably announce the unicode character name which isn't super helpful

File: frontend/src/Components/SearchBar/SearchBar.jsx [hunk 6]
```diff
@@ -16,9 +17,17 @@ const SearchBar = ({ onSelect, onSearchOnMap }) => {
...
+    }
+
+    const controller = new AbortController();
+    latestRequest.current = controller;
+
     try {
       const response = await fetch(
-        `http://localhost:3000/api/businesses/search?query=${value}`
+        `http://localhost:3000/api/businesses/search?query=${value}`,
+        { signal: controller.signal }
```
Comment (sinap-fb) [inline frontend/src/Components/SearchBar/SearchBar.jsx:30]: nicely done!

File: backend/dataPreprocessor.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    item.rated = item.rated !== null ? item.rated : 0;
+    item.reviewed = item.reviewed !== null ? item.reviewed : "";
+    return item;
+  });
+};
+
+// Normalize text data
+const normalizeTextData = (data) => {
+  return data.map((item) => {
+    item.reviewed = item.reviewed.toLowerCase().replace(/[^a-z0-9\s]/g, "");
```
Comment (mylesdomingo) [inline backend/dataPreprocessor.js:31]: Most people don't know regex; I assume this is sourced from somewhere, can you add what the query does in the comments for visibilty? Looks like it checks for alphanumerics.

File: backend/dataPreprocessor.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      User: true,
+      Business: true,
+    },
+  });
+};
+
+// Handle missing values
+const handleMissingValues = (data) => {
+  return data.map((item) => {
+    item.rated = item.rated !== null ? item.rated : 0;
```
Comment (mylesdomingo) [inline backend/dataPreprocessor.js:22]: Do we want missing values to be handled this way? I think this was a similar situation with your first project where null reviews were rated as 0. 

Generally, I think we should avoid using any non-standard data and filter out data that have missing values.

File: backend/index.js [hunk 9]
```diff
@@ -193,8 +229,63 @@ app.get("/api/categories", async (req, res) => {
   }
 });
 
-// Process JSON file
-processJsonFile();
+//Interact route
```
Comment (sinap-fb) [inline backend/index.js:232]: appreciate the comment! maybe a bit more clarity would be even more helpful, something like "Interact route - used for various interactions between users and businesses"

File: backend/dataPreprocessor.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+// TF-IDF encode text data
+const tfidfEncodeTextData = (data) => {
+  data.forEach((item) => {
+    tfidf.addDocument(item.reviewed);
+  });
+
+  return data.map((item, index) => {
+    const tfidfVector = {};
```
Comment (mylesdomingo) [inline backend/dataPreprocessor.js:53]: Can you add a little more context here on TF-IDF encoding and what it's meant to achieve in layman terms?

File: backend/index.js [hunk 11]
```diff
@@ -193,8 +229,63 @@ app.get("/api/categories", async (req, res) => {
...
+  } catch (error) {
+    console.error("Error fetching favorites:", error);
+    res.status(500).json({ error: "Failed to fetch favorites" });
+  } finally {
+    await prisma.$disconnect();
+  }
+});
+
+// Endpoint to trigger data preprocessing
+app.get("/preprocess-data", (req, res) => {
```
Comment (sinap-fb) [inline backend/index.js:275]: Just as a heasd up: if this is an expensive operation, you probably wouldn't want to make this publicly accessible (or you might add rate limiting) since malicious users could abuse it.  (E.g. create a ton of OS processes via the `exec()`).  But I presume these sort of security things are not in the scope of the course.

File: backend/jsonProcessor.js [hunk 12]
```diff
@@ -67,11 +66,10 @@ async function processJsonFile() {
             placeId: business.place_id,
             name: placeDetails.name || business.name,
             location,
-            contactInformation:
-              placeDetails.formatted_phone_number || "Unknown",
+            contactInformation: placeDetails.formatted_phone_number || null,
             overview: placeDetails.editorial_summary
               ? placeDetails.editorial_summary.overview
-              : "Overview not provided",
+              : null,
```
Comment (mylesdomingo) [inline backend/jsonProcessor.js:72]: In the future for cleaner commits I'd add changes requested to the same PR; this is fine for now.

File: backend/trainModel.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  // Defining the model
+  const model = tf.sequential();
+  model.add(
+    tf.layers.dense({
+      units: 10,
+      activation: "relu",
+      inputShape: [inputTensor.shape[1]],
+    })
+  );
+  model.add(tf.layers.dense({ units: 1, activation: "linear" }));
```
Comment (mylesdomingo) [inline backend/trainModel.js:41]: We're using linear regression model here right? Can you also clarify in layman terms what you're aiming to use machine learning for?

Overall it looks semantically correct; just wanted to check that we're training on the right feature data.

File: frontend/src/Components/BusinessCard/BusinessCard.css [hunk 14]
```diff
@@ -0,0 +1,63 @@
+.business-card {
+  background-color: white;
+  border-radius: 10px;
+  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
+  overflow: hidden;
+  cursor: pointer;
```
Comment (sinap-fb) [inline frontend/src/Components/BusinessCard/BusinessCard.css:6]: using cursor: pointer doesn't really make sense to me since the card isn't clickable (only a few things *within* the card are clickable)

File: frontend/src/Components/FavoriteBusinesses/FavoriteBusinesses.jsx [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+  const [selectedCategory, setSelectedCategory] = useState("");
+  const user = JSON.parse(localStorage.getItem("user"));
+  const [businesses, setBusinesses] = useState([]);
+  const [groupedBusinesses, setGroupedBusinesses] = useState({});
+  const apiKey = import.meta.env.VITE_API_KEY;
+
+  const groupBusinessesByCategory = (businesses) => {
+    const grouped = businesses.reduce((acc, business) => {
+      const category = business.category || "Uncategorized";
+      if (!acc[category]) {
```
Comment (mylesdomingo) [inline frontend/src/Components/FavoriteBusinesses/FavoriteBusinesses.jsx:17]: You know I thought this reduce function wouldn't work at first but I saw the initial value for acc set as `{}`. 
Personally I would use a map but this works too. Nice job!

File: frontend/src/Components/FavoriteBusinesses/FavoriteBusinesses.jsx [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+const FavoriteBusinesses = () => {
+  const [selectedCategory, setSelectedCategory] = useState("");
+  const user = JSON.parse(localStorage.getItem("user"));
+  const [businesses, setBusinesses] = useState([]);
+  const [groupedBusinesses, setGroupedBusinesses] = useState({});
+  const apiKey = import.meta.env.VITE_API_KEY;
+
+  const groupBusinessesByCategory = (businesses) => {
+    const grouped = businesses.reduce((acc, business) => {
+      const category = business.category || "Uncategorized";
```
Comment (mylesdomingo) [inline frontend/src/Components/FavoriteBusinesses/FavoriteBusinesses.jsx:16]: I'd use a null coalescer to return a fallback category;  I think this wouolod return a boolean instead. 
`const category = business.category ?? "Uncategorized"`

File: frontend/src/Components/FavoriteBusinesses/FavoriteBusinesses.jsx (hunk 15 above)
Comment (mylesdomingo) [reply frontend/src/Components/FavoriteBusinesses/FavoriteBusinesses.jsx:17]: Although I am a bit surprised we don't preprocess this data beforehand. I feel like prisma maybe has some optimizations for this use case. I don't think it would impact performance too much currently. Maybe a task for later?

File: frontend/src/Components/ReviewForm/ReviewForm.jsx [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+      });
+
+      if (!response.ok) {
+        throw new Error(`HTTP error! Status: ${response.status}`);
+      }
+
+      const savedReview = await response.json();
+      onReviewSubmit(savedReview);
+      handleReviewAndRating(businessId);
+      setReviewText("");
```
Comment (mylesdomingo) [inline frontend/src/Components/ReviewForm/ReviewForm.jsx:35]: I'd put everything into onReviewSubmit tbh:

handleReviewAndRating(businessId);
setReviewText("")
setRating(0);

**Review by sinap-fb — COMMENTED** [review]
good work!

File: backend/dataPreprocessor.js (hunk 8 above)
Comment (jose-esquivel) [reply backend/dataPreprocessor.js:22]: I agree with this, you might get unexpected results if you treat non-rated items as 0.

File: backend/index.js (hunk 11 above)
Comment (jose-esquivel) [reply backend/index.js:275]: Agree, doesn't seem like this should be a public endpoint. Probably better to make a command in your package.json, if you need to run this frequently. eg. "npm run data-processor".
